[Feature] PhoneAsItem Support

ItsANoBrainer opened the pull request:
- Added support for PhoneAsItem export from NPWD
- Added config with array of acceptable phone items in case a server has multiple phone types

Users will still need to enable this feature in the NPWD config.json.
- Set "enabled" to true
- Set "exportResource" to "qb-npwd"
- Set "exportFunction" to "HasPhone"

ItsANoBrainer committed:
[Feature] PhoneAsItem Support

- Added support for PhoneAsItem export from NPWD

BerkieBb commented:
1. Use the hasitem callback instead of making your own
2. Use promises like in the hasitem function or use the hasitem function of qb-core

ItsANoBrainer committed:
Updated implementation

- Using QBCore:HasItem callback instead of custom one

ItsANoBrainer committed:
Promise implementation

Now uses promises instead of silly loop!

BerkieBb commented:
And change the Config.PhoneList to
```lua
Config.PhoneList = {
    ['phone'] = 1,
}
```
To accompany for the supported format for the hasitem callback

ItsANoBrainer committed:
Config format change

ItsANoBrainer committed:
Updated HasPhone

Now uses QBCore.Functions.HasItem, which does the exact same thing I did with the promises. No reason to reinvent the wheel. I did not know this was a thing beforehand.

ItsANoBrainer committed:
Refactor

Now uses the PlayerData inventory state to manage if the player has a phone or not, as well as if it should be disabled or not via the export.

ItsANoBrainer committed:
Feature update + README update

- Updated README for instructions
Will no longer DoPhoneCheck on resource start if npwd isnt started.
- Phone items in the config are now registered as usable items to open the phone ui
- Updated manifest to reflect changes

IdrisDose commented:
Code looks alrighty, will need testing before merging

TasoOneAsia commented on client.lua:
```diff
@@ -0,0 +1,50 @@
+local QBCore = exports['qb-core']:GetCoreObject()
+local hasPhone = false
+
+local function DoPhoneCheck(PlayerItems)
+    hasPhone = false
```
This pattern has a code smell, I can feel it in the force, beware when refactoring in the future.

IdrisDose replied:
Yeah I agree, not the greatest way.

ItsANoBrainer replied:
Suggestions?

RockySouthpaw replied:
> This pattern has a code smell, I can feel it in the force, beware when refactoring in the future.

If you have no suggestions to improve the smell, please cover your nose sir

TasoOneAsia commented on client.lua:
```diff
@@ -0,0 +1,50 @@
+local QBCore = exports['qb-core']:GetCoreObject()
+local hasPhone = false
+
+local function DoPhoneCheck(PlayerItems)
+    hasPhone = false
+    
+    if PlayerItems then
+        for _,item in pairs(PlayerItems) do
+            if Config.PhoneList[item.name] then
+                hasPhone = true
+                break;
+            end
+        end
+    end
+
+    exports['npwd']:setPhoneDisabled(not hasPhone)
+end
```
Small indentation thing

```suggestion
local function DoPhoneCheck(PlayerItems)
    hasPhone = false

    if not PlayerItems then return end

    for _,item in pairs(PlayerItems) do
        if Config.PhoneList[item.name] then
            hasPhone = true
            break;
        end
    end

    exports['npwd']:setPhoneDisabled(not hasPhone)
end
```

ItsANoBrainer replied:
Done

TasoOneAsia reviewed (changes requested):
Generally looks LGTM, has a smell that could be a problem in the future if complexity in this resource increases but I am fine with that for now.

Literally just the indentation, is all.

ItsANoBrainer committed:
Update client.lua

Co-authored-by: Taso <taso@projecterror.dev>
